external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/InternalKafkaConsumerPool.scala

```diff
@@ -0,0 +1,241 @@
+
+  def getNumActive: Int = pool.getNumActive
+
+  def getNumActive(key: CacheKey): Int = pool.getNumActive(key)
+
+  def getTotal: Int = getNumIdle + getNumActive
+
+  def getTotal(key: CacheKey): Int = getNumIdle(key) + getNumActive(key)
+
+  private def updateKafkaParamForKey(key: CacheKey, kafkaParams: ju.Map[String, Object]): Unit = {
+    // We can assume that kafkaParam should not be different for same cache key,
+    // otherwise we can't reuse the cached object and cache key should contain kafkaParam.
+    // So it should be safe to put the key/value pair only when the key doesn't exist.
+    objectFactory.keyToKafkaParams.putIfAbsent(key, kafkaParams)
+  }
+
+  private def extractCacheKey(consumer: InternalKafkaConsumer): CacheKey = {
+    new CacheKey(consumer.topicPartition, consumer.kafkaParams)
+  }
+}
+
+private[kafka010] object InternalKafkaConsumerPool {
+
+  /**
+   * Builds the pool for [[InternalKafkaConsumer]]. The pool instance is created per each call.
+   */
+  def build: InternalKafkaConsumerPool = {
+    val objFactory = new ObjectFactory
+    val poolConfig = new PoolConfig
+    new InternalKafkaConsumerPool(objFactory, poolConfig)
+  }
+
+  case class PooledObjectInvalidated(key: CacheKey, lastInvalidatedTimestamp: Long,
+                                     lastBorrowedTime: Long) extends RuntimeException
+
+  object CustomSwallowedExceptionListener extends SwallowedExceptionListener with Logging {
+    override def onSwallowException(e: Exception): Unit = {
+      logError(s"Error closing Kafka consumer", e)
+    }
+  }
+
+  class PoolConfig extends GenericKeyedObjectPoolConfig[InternalKafkaConsumer] {
+    private var softMaxTotal = Int.MaxValue
+
+    def getSoftMaxTotal(): Int = softMaxTotal
+
+    init()
+
+    def init(): Unit = {
+      import PoolConfig._
+
+      val conf = SparkEnv.get.conf
+
+      softMaxTotal = conf.getInt(CONFIG_NAME_CAPACITY, DEFAULT_VALUE_CAPACITY)
+
+      val jmxEnabled = conf.getBoolean(CONFIG_NAME_JMX_ENABLED,
+        defaultValue = DEFAULT_VALUE_JMX_ENABLED)
+      val minEvictableIdleTimeMillis = conf.getLong(CONFIG_NAME_MIN_EVICTABLE_IDLE_TIME_MILLIS,
+        DEFAULT_VALUE_MIN_EVICTABLE_IDLE_TIME_MILLIS)
+      val evictorThreadRunIntervalMillis = conf.getLong(
+        CONFIG_NAME_EVICTOR_THREAD_RUN_INTERVAL_MILLIS,
+        DEFAULT_VALUE_EVICTOR_THREAD_RUN_INTERVAL_MILLIS)
+
+      // NOTE: Below lines define the behavior, so do not modify unless you know what you are
+      // doing, and update the class doc accordingly if necessary when you modify.
+
+      // 1. Set min idle objects per key to 0 to avoid creating unnecessary object.
+      // 2. Set max idle objects per key to 3 but set total objects per key to infinite
+      // which ensures borrowing per key is not restricted.
+      // 3. Set max total objects to infinite which ensures all objects are managed in this pool.
+      setMinIdlePerKey(0)
+      setMaxIdlePerKey(3)
+      setMaxTotalPerKey(-1)
```

Comment:
This is intended to be unbounded, because existing behavior doesn't block borrowing consumers. It just used non-cached consumers instead, and it always succeeded so we should ensure this behavior too. 
IMHO min idle/max idle is the key to maintain optimal pool, so it may be good for expert to enable configuration on these values for tuning pool.

Reply:
I know it's the same in the original implementation but this is vulnerable to short time resource allocation flood which can't be handled with idle time.

Reply:
Hmm... yeah it would make sense, though it still needs to provide consumers whenever tasks request. 

Two approaches come in my mind: 1. bounded & provide non-cached consumer with some delays when borrow fails 2. unbounded & apply random delay when borrowing. Former requires me to reintroduce handling of cached vs non-cached so would discard simplicity on my patch, whereas latter just requires me to add random delay but the chance to avoid resource allocation flood would be completely based on probability. (max ~50 ms might be enough?).

I'd like to lean on 2 but also open to 1 if we don't feel safe with 2.

Reply:
Yeah, the first approach is heavy and doesn't solve the problem itself. The second case I'm not sure how this can help (maybe not understood properly). Additionally it could introduce fluctuation in the performance. Maybe we can leave the soft boundary (as it's like that in the original implementation and I don't see jiras with it) and just put warning on each new allocation after a threshold. This way admins/devs can see this in the log and don't have to monitor the cache constantly.

Reply:
Actually soft boundary with existing cache is more like total keys in new cache (and new cache implements it), while you're concerning about it is regarding same key having too many tasks hence too many consumers.

In new cache we have more metrics: the number of idle/active consumers for given key, the number of total idle/active consumers. We can check the metrics and log when it exceeds the threshold or meet condition. Even simpler, we can just leave log message whenever cache **creates** consumer instead of returning idle consumer, if it would help end users to tune pool as well as tasks.

Reply:
In my view the mentioned simple debug log message is fine. This can be seen even without constant monitoring.

Reply:
Let's see how committers think about it. We can wait and react.